admin/services/restart.service.js

```diff
@@ -16,24 +24,81 @@ restartService.getPupils = async (schoolId) => {
   const school = await schoolDataService.findOne({_id: schoolId})
   if (!school) throw new Error(`School [${schoolId}] not found`)
   let pupils = await pupilDataService.getSortedPupils(schoolId, 'lastName', 'asc')
-  pupils = pupils
-    .filter(p => restartService.isPupilEligible(p))
-    .map(({ _id, pin, dob, foreName, middleNames, lastName }) =>
+  pupils = await Promise.filter(pupils.map(async p => {
+    const isPupilEligible = await restartService.isPupilEligible(p)
+    if (isPupilEligible) return p
+  }), p => !!p)
+  if (pupils.length === 0) return []
+  pupils = pupils.map(({ _id, pin, dob, foreName, middleNames, lastName }) =>
       ({ _id, pin, dob: () => dateService.formatLongGdsDate(dob), foreName, middleNames, lastName }))
   pupils = pupilIdentificationFlagService.addIdentificationFlags(pupils)
   return pupils
 }
 
 /**
- * Determine if pupil is eligible for restart
+ * Determine all the factors that make pupil eligible for a restart
  * @param p
  * @returns {boolean}
  */
-restartService.isPupilEligible = (p) => {
-  // TODO: As part of the next PR pupil restart collection is introduced which will determine the number of restarts per pupil
-  if (p.attendanceCode) return false
+restartService.isPupilEligible = async (p) => {
+  const canRestart = await restartService.canRestart(p._id)
+  if (p.attendanceCode || !canRestart) return false
   const hasExpiredToday = p.pinExpiresAt && moment(p.pinExpiresAt).isSame(moment.now(), 'day')
   return !pinValidator.isActivePin(p.pin, p.pinExpiresAt) && hasExpiredToday
 }
 
+/**
+ * Perform restart for a list of pupils
+ * @param pupilsList
+ * @param restartReason
+ * @param didNotCompleteInfo
+ * @param restartFurtherInfo
+ * @param userId
+ * @returns {Promise.<void>}
+ */
+
+restartService.restart = async (pupilsList, restartReason, didNotCompleteInfo, restartFurtherInfo, userName) => {
+  await pinService.expireMultiplePins(pupilsList)
+  // All pupils should be eligible for restart before proceeding with creating a restart record for each one
+  const eligibilityList = await Promise.all(pupilsList.map(async pupilId => {
```

Comment:
It might be useful if this functionality was extracted into it's own function: `canAllPupilsRestart(pupilList)` returning bool.

Reply:
@jon-shipley updated